consensus/src/common.rs

```diff
@@ -242,7 +244,77 @@ impl Consensus for ProtocolStateChain {
             Ok(a_prev_lock_checkpoint == b_prev_lock_checkpoint)
         } else {
             // Check for previous epoch case using both orientations
-            Ok(check(&a, &b, candidate.epoch_slot()) || check(&b, &a, self.epoch_slot()))
+            Ok(check(a, b, candidate.epoch_slot()) || check(b, a, self.epoch_slot()))
+        }
+    }
+
+    fn config(&self) -> ConsensusConstants {
+        ConsensusConstants::mainnet()
+    }
+
+    /// Computes the relative minimum window density of the given chains.
+    /// The minimum density value is used in the case of a long range fork
+    /// and the chain with the higher minimum window density is chosen as the canonical chain.
+    /// The need for relative density is explained here:
+    /// <https://github.com/MinaProtocol/mina/blob/02dfc3ff0160ba3c1bbc732baa07502fe4312b04/docs/specs/consensus/README.md#5412-relative-minimum-window-density>
+    fn relative_min_window_density(
+        &self,
+        chain_b: &ProtocolStateChain,
+    ) -> Result<u32, ConsensusError> {
+        let tip_state = self
+            .consensus_state()
+            .ok_or(ConsensusError::ConsensusStateNotFound)?;
+        let chain_b = chain_b
+            .consensus_state()
+            .ok_or(ConsensusError::ConsensusStateNotFound)?;
+
+        // helpers for readability.
+        let min = |a: u32, b: u32| a.min(b);
+        let max = |a: u32, b: u32| a.max(b);
+
+        let max_slot = max(
+            tip_state.curr_global_slot.slot_number.0,
+            chain_b.curr_global_slot.slot_number.0,
+        );
+
+        // grace-period rule
+        if max_slot < self.config().grace_period_end.0 {
+            return Ok(tip_state.min_window_density.0);
         }
+
+        let projected_window = {
+            // compute shift count
+            let mut shift_count = min(
+                max(
+                    max_slot - tip_state.curr_global_slot.slot_number.0.saturating_sub(1),
+                    0,
+                ),
+                self.config().sub_windows_per_window.0,
+            );
+            // initialize projected window based off of chain_a
+            let mut projected_window = tip_state.sub_window_densities.clone();
+
+            // relative sub window
+            let mut rel_sub_window =
+                tip_state.curr_global_slot.slot_number.0 % self.config().sub_windows_per_window.0;
+
+            // ring shift
+            while shift_count > 0 {
+                rel_sub_window = (rel_sub_window + 1) % self.config().sub_windows_per_window.0;
+                projected_window[rel_sub_window as usize] = Length(0);
```

Comment:
I know this matches OCaml implementation, but calling this a `Length` is confusing.  It would be nicer to create a type called `Density` or to keep it simple, call it whatever it is.. e.g. `u32` or `u64`


Actually... on second thought, it probably ties back to serialisation and therefore would be a pain to change it and make things confusing.

So maybe because we want this to map to real mina  pub/network protocol types it would be too confusing to change it like I suggested earlier.

Unless the serialisation layout stuff had a way to map things to different names automatically.. but again, sounds complicated.

Thus, I retract my previous uninformed suggestion.  Instead maybe just put a comment next to it like `// really a density` or maybe you can just create a local alias!

Reply:
Yeah, `Length` is somewhat confusing, but it's there because it had to match mina's serialization types. I've just added a comment saying `// density at rel_sub_window index`